src/util.cc

```diff
@@ -218,6 +221,51 @@ int WriteFileSync(v8::Isolate* isolate,
   return WriteFileSync(path, buf);
 }
 
+Maybe<std::string> ReadFileSync(const std::string& path) {
+  if (strlen(path.c_str()) != path.length()) {
+    return Nothing<std::string>();
+  }
+
+  uv_fs_t req;
+
+  const int fd = uv_fs_open(nullptr, &req, path.c_str(), O_RDONLY, 0, nullptr);
+  uv_fs_req_cleanup(&req);
+  if (fd < 0) {
+    return Nothing<std::string>();
+  }
+
+  std::shared_ptr<void> defer_close(nullptr, [fd] (...) {
+    uv_fs_t close_req;
+    CHECK_EQ(0, uv_fs_close(nullptr, &close_req, fd, nullptr));
+    uv_fs_req_cleanup(&close_req);
+  });
+
+  const size_t kBlockSize = 32 << 10;
+  std::string contents;
+  size_t offset = 0;
+  while (true) {
+    contents.resize(offset + kBlockSize);
+
+    uv_buf_t buf = uv_buf_init(&contents[offset], kBlockSize);
+
+    const int r = uv_fs_read(nullptr, &req, fd, &buf, 1, offset, nullptr);
+    uv_fs_req_cleanup(&req);
+
+    if (r < 0) {
+      return Nothing<std::string>();
+    }
+
+    offset += r;
+
+    if (r != kBlockSize) {
+      break;
+    }
+  }
+  contents.resize(offset);
```

Comment:
So, `InternalModuleReadJSON()` was deliberately structured to avoid the final resize because 9 out of 10 times the file is only a fraction of the 32k buffer, and shrinking it will result in a realloc and the concomitant copy.

I suggest changing the function prototype and moving the resize logic to the call sites. That makes it a little more cumbersome to use but that seems like an okay trade-off to me. Performance matters here.

Suggested prototype:
```c++
// Passing path as pointer-to-char to avoid unnecessary c-string-to-std-string conversion
ssize_t ReadFileSync(const char* path, size_t pathlen, std::string* contents) {
  // ...
}
```
And call it like this:
```c++
node::Utf8Value path(isolate, args[0]);
std::string contents;
ssize_t nread = ReadFileSync(*path, path.length(), &contents);
if (nread < 0) {
  // handle error
} else {
  contents.resize(static_cast<size_t>(nread));
  // use contents
}
```
The nul byte check becomes:
```c++
if (strlen(path) != pathlen)
  return 0;
```

Reply:
should that be return -1 or something? I assume we want to distinguish between invalid filename and successful read of empty file?

Reply:
That's fine too, it could be `return UV_EINVAL;`.